### PR TITLE
Fix security issues identified by bandit tool

### PR DESCRIPTION
**Pull request by tjtg**
This PR contains fixes for a variety of security issues identified using the `bandit` tool. The issues identified are generally low risk, but it is preferable to avoid the issues entirely where possible.

- Replace subprocess calls to `rm` and `touch` with Python standard library functions.
- Avoid usage of assertions in non-test code as these get disabled in optimised bytecode.
- Mark usage of pickle and md5 as `nosec` as these are used for hash-based comparison of objects and not for security/cryptography purposes.
- Change random generation to use numpy function to clarify the purpose is not security/cryptography.
- Remove usage of `eval()` and the `safe_eval` wrapper function.

Coordinate system options are passed to the `neighbour-finding` CLI using a JSON string rather than a Python code snippet to be eval'd. This requires accompanying changes to the CLI tests.

There is another use of `eval()` remaining in the weather symbols code. That one is harder to remove, will require further thought and hasn't been addressed in this branch.

Testing:
 - [x] Ran tests and they passed OK

### CODE REVIEW

File: lib/improver/cli/neighbour_finding.py [hunk 1]
```diff
@@ -276,8 +285,18 @@ def process(orography, landmask, site_list, metadata_dict=None,
...
+        site_crs = getattr(ccrs, scrs)
+        scrs_opts = json.loads(kwargs['site_coordinate_options'])
+        if 'globe' in scrs_opts:
+            crs_globe = ccrs.Globe(**scrs_opts['globe'])
+            del scrs_opts['globe']
+        else:
+            crs_globe = ccrs.Globe()
+        kwargs['site_coordinate_system'] = site_crs(
+            globe=crs_globe, **scrs_opts)
+        del kwargs['site_coordinate_options']
```
Comment (arh89) [inline lib/improver/cli/neighbour_finding.py]: Is there a reason you didn't `pop` this on L291? (I can understand the explicit `del scts_opts['globe']`)

File: lib/improver/tests/utilities/test_load.py [hunk 2]
```diff
@@ -249,10 +248,7 @@ def setUp(self):
 
     def tearDown(self):
         """Remove temporary directories created for testing."""
-        Call(['rm', '-f', self.filepath])
-        Call(['rm', '-f', self.low_cloud_filepath])
-        Call(['rm', '-f', self.med_cloud_filepath])
-        Call(['rmdir', self.directory])
+        shutil.rmtree(self.directory, ignore_errors=True)
```
Comment (carolinejjones) [inline lib/improver/tests/utilities/test_load.py]: Not sure about this change. shutil.rmtree(self.directory, ignore_errors=True) will try and delete everything in this directory and below. What if self.directory has been set to something stupid. The previous code only tried to remove the files it think it has created and the directory if it is empty.

Comment (tjtg) [reply lib/improver/tests/utilities/test_load.py]: I've changed this and other similar code to use `os.remove` and `os.rmtree` functions to only remove specified files/directories.
Some of these file removals need to be wrapped in `try`/`except FileNotFoundError`/`pass` as the teardown function is shared across a group of tests where some tests create the specified files and others do not.

File: lib/improver/cli/neighbour_finding.py (hunk 1 above)
Comment (tjtg) [reply lib/improver/cli/neighbour_finding.py]: No reason other than not thinking of it when I was writing the code. I've changed this to use `pop` as it's a neater way to write the same thing.

File: lib/improver/tests/utilities/test_load.py (hunk 2 above)
Comment (arh89) [reply lib/improver/tests/utilities/test_load.py]: @carolinejjones - if you're happy with this change, can you hit merge please?

Comment (carolinejjones) [reply lib/improver/tests/utilities/test_load.py]: I'm happy with these changes. I've checked with Aaron and these changes have been tested. I shall hit the merge button

**Pull request merged**